### PR TITLE
SJQ2T-226 BE - Endpoint for Edit Bucket

### PR DESCRIPTION
**Pull request by MarkoRadunovicHtec**
BucketEditRequest added, Endpoint for edit bucket functionality added, Tests added, **Custom Validation classes renamed** and added new one for edit Bucket. Test for Edit Bucket added. Need to add test for validator class and add check if Bucket that needs to be edited is not already deleted

### CODE REVIEW

File: backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import javax.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+@Constraint(validatedBy = EditBucketValidator.class)
+public @interface CanBucketBeEdited {
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java]: UserCanEditBucket? CurrentUserCanEditBucket?

File: backend/src/main/java/com/htecgroup/skynest/annotation/validator/EditBucketValidator.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  private final BucketService bucketService;
+
+  @Override
+  public boolean isValid(UUID uuid, ConstraintValidatorContext context) {
+    LoggedUserDto loggedUserDto = currentUserService.getLoggedUser();
+    UUID loggedUserUuid = loggedUserDto.getUuid();
+
+    UUID ownerId = bucketService.getBucket(uuid).getCreatedById();
+
+    if (!loggedUserUuid.equals(ownerId)) {
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/annotation/validator/EditBucketValidator.java:27]: and if the bucket is private?

File: backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java [hunk 3]
```diff
@@ -174,4 +169,13 @@ public ResponseEntity<BucketResponse> getBucket(@PathVariable UUID uuid) {
         new ResponseEntity<>(bucketResponse, HttpStatus.OK);
     return bucketResponseEntity;
   }
+
+  @PutMapping("/{uuid}")
+  public ResponseEntity<BucketResponse> editBucket(
+      @Valid @RequestBody BucketEditRequest bucketEditRequest, @PathVariable UUID uuid) {
+    ResponseEntity<BucketResponse> bucketResponseEntity =
+        new ResponseEntity<>(bucketService.editBucket(bucketEditRequest, uuid), HttpStatus.OK);
+    log.info("Bucket is successfully edited");
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java]: should our logs be a bit more detailed? something like:

User <email> (<id>) edited bucket <name> (<id>)

I've already done something like that with [actions](https://github.com/JDP2022SkyNest/SkyNest/pull/114/files#diff-a9284f5ec0b2eca63667d4d3273a734549289b982cca43bac369e4bdc31d3120), so you could use that when it's merged

File: backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  @NotBlank(message = "cannot be null or empty")
+  @Size(max = 1000, message = "length cannot be over 1000 characters")
+  private String name;
+
+  @NotBlank(message = "cannot be null or empty")
+  @Size(max = 5000, message = "length cannot be over 5000 characters")
+  private String description;
+
+  private String isPublic;
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java]: no restrictions?

probably fine if this a Boolean/boolean

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java [hunk 5]
```diff
@@ -14,10 +16,13 @@
 import lombok.AllArgsConstructor;
 import org.modelmapper.ModelMapper;
 import org.springframework.stereotype.Service;
+import org.springframework.validation.annotation.Validated;
 
+import javax.validation.Valid;
 import java.util.UUID;
 
 @Service
+@Validated
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java:33]: we need both this and Valid for each method param?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java [hunk 6]
```diff
@@ -53,4 +58,16 @@ public BucketResponse getBucket(UUID uuid) {
...
   }
+
+  @Override
+  public BucketResponse editBucket(
+      BucketEditRequest bucketEditRequest, @Valid @CanBucketBeEdited UUID uuid) {
+    BucketEntity bucketEntity =
+        bucketRepository.findById(uuid).orElseThrow(BucketNotFoundException::new);
+    bucketEditRequest.setName(bucketEditRequest.getName().trim());
+    bucketEditRequest.setDescription(bucketEditRequest.getDescription().trim());
+    modelMapper.map(bucketEditRequest, bucketEntity);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java:121]: setPublic?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java (hunk 5 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java:33]: If its custom annotation, like in this case than yes.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java (hunk 6 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java:121]: no need, modelMapper does the job

File: backend/src/main/java/com/htecgroup/skynest/annotation/validator/EditBucketValidator.java (hunk 2 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/annotation/validator/EditBucketValidator.java:27]: Only owner can edit it, no matter is it private or public. I think thats what client said. For now owner is only creator.

File: backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java (hunk 4 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java]: Fixed. There are restrictions now.

File: backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java [hunk 7]
```diff
@@ -294,4 +295,80 @@ public ResponseEntity<Boolean> deleteBucket(@PathVariable UUID uuid) {
...
+                                + " \"timestamp\": \"2022-06-07 16:18:12\"}")
+                  })
+            })
+      })
+  @PutMapping("/{uuid}")
+  public ResponseEntity<BucketResponse> editBucket(
+      @Valid @RequestBody BucketEditRequest bucketEditRequest, @PathVariable UUID uuid) {
+    ResponseEntity<BucketResponse> bucketResponseEntity =
+        new ResponseEntity<>(bucketService.editBucket(bucketEditRequest, uuid), HttpStatus.OK);
+    log.info("Bucket ({}) was successfully edited", bucketResponseEntity.getBody().getBucketId());
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java]: use actions

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java [hunk 8]
```diff
@@ -82,4 +87,20 @@ public void deleteBucket(UUID uuid) {
...
+    BucketEntity bucketEntity =
+        bucketRepository.findById(uuid).orElseThrow(BucketNotFoundException::new);
+    BucketDto bucketDto = findBucketById(uuid);
+    if (bucketDto.getDeletedOn() != null) {
+      throw new BucketAlreadyDeletedException();
+    }
+    bucketEditRequest.setName(bucketEditRequest.getName().trim());
+    bucketEditRequest.setDescription(bucketEditRequest.getDescription().trim());
+    modelMapper.map(bucketEditRequest, bucketEntity);
+    bucketRepository.save(bucketEntity);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java]: record this action after this line

File: backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java (hunk 1 above)
Comment (ivanfhtec) [reply backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java]: any suggestions?

File: backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java (hunk 4 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/model/request/BucketEditRequest.java]: When this @Pattern is applied it already has to be true or false, Really should be a Boolean

File: backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java (hunk 1 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/annotation/CanBucketBeEdited.java]: Changed to CurrentUserCanEditBucket, gives most info.

File: backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java (hunk 3 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/controller/BucketController.java]: Fixed it to be more detailed.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java (hunk 8 above)
Comment (MarkoRadunovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java]: recorded

**Review by ivanfhtec — APPROVED** [review]

**Pull request merged**